feat(kafka): add aws_msk_single_scram_secret_association resource

Brightside56 opened the pull request:
<!--
Please read through https://git.io/fj2m9 if this is your first time opening a
pull request to this repo. Find us in https://crossplane.slack.com
if you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your
reviewers' attention to anything that needs special consideration.

We love pull requests that resolve an open issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":
-->

adds [aws_msk_single_scram_secret_association](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/msk_single_scram_secret_association) resource

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested
Provider was built and manually installed to development cluster. I've tested creation and deletion of added resource
<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

jeanduplessis commented:
/test-examples="examples/kafka/v1beta1/singlescramsecretassociation.yaml"

Brightside56 commented:
@jeanduplessis I see, will take a look and try to fix soon

mbbush commented on config/kafka/config.go:
```diff
@@ -233,6 +233,14 @@ func Configure(p *config.Provider) { //nolint:gocyclo
+	p.AddResourceConfigurator("aws_msk_single_scram_secret_association", func(r *config.Resource) {
+		r.References["secret_arn"] = config.Reference{
+			TerraformName:     "aws_secretsmanager_secret",
+			RefFieldName:      "SecretArnRef",
+			SelectorFieldName: "SecretArnSelector",
```
Better to leave these unset unless there's a need to change the default name (such as when the generator pluralizes incorrectly)

Brightside56 replied:
Okay, how can I check if it pluralizes incorrectly?

Brightside56 replied:
@mbbush https://github.com/crossplane-contrib/provider-upjet-aws/pull/1754/commits/9861ed5f18c967d5eb9bb0ed96c931cca1e7f2dd - with unset reference

mbbush replied:
That's exactly what I had in mind. Thanks!

mbbush commented on examples/kafka/v1beta1/singlescramsecretassociation.yaml:
```diff
@@ -0,0 +1,97 @@
+  annotations:
+    meta.upbound.io/example-id: kafka/v1beta1/scramsecretassociation
+  labels:
+    testing.upbound.io/example-name: example
```
Could you give this a more specific name, and then keep it consistent for all the resources?

Brightside56 replied:
Sure

But what is a difference between examples and examples-generated? Where example for marketplace is taken from? (in case of scramsecretassociation it contains single resource, however examples and examples-generated of scramsecretassociation include dependencies)

mbbush replied:
`examples-generated` is generated by upjet based on the examples in the terraform provider documentation. This is also how upjet knows to automatically generate certain reference fields. Sometimes the examples are incomplete, or contain HCL that the codegen tool can't handle, so they're generally used as a starting place for the handwritten `examples`, which are what actually get published to the upbound marketplace, and also used as manifests in uptest.

mbbush commented on examples/kafka/v1beta1/singlescramsecretassociation.yaml:
```diff
@@ -0,0 +1,97 @@
+        testing.upbound.io/example-name: example
+---
+apiVersion: kms.aws.upbound.io/v1beta1
+kind: Key
```
IIRC MSK also requires a key policy to allow it to use this key

mbbush commented on examples/kafka/v1beta1/singlescramsecretassociation.yaml:
```diff
@@ -0,0 +1,97 @@
+  name: example
+spec:
+  forProvider:
+    policy: ${data.aws_iam_policy_document.example.json}
```
You'll need to actually put the json here

mbbush reviewed (commented):
Generally looking good, just needs some minor work on the example/test. 

I wonder if it's worth adding documentation to either this resource or the existing scram secrets association resource indicating that they conflict with each other, or whether that's already obvious. I think I'm leaning towards "that's already obvious", especially since the existing resource has a clear warning that it manages *all* secrets, but I'd be happy to hear someone else's opinion.

Brightside56 commented:
> Generally looking good, just needs some minor work on the example/test.
> 
> I wonder if it's worth adding documentation to either this resource or the existing scram secrets association resource indicating that they conflict with each other, or whether that's already obvious. I think I'm leaning towards "that's already obvious", especially since the existing resource has a clear warning that it manages _all_ secrets, but I'd be happy to hear someone else's opinion.

I've replaced examples with examples from ScramSecretAssociation (with minor edits, I think it should work), added note

Brightside56 commented:
@jeanduplessis @mbbush can I ask you run tests at least to check if they're passing?

turkenf commented:
/test-examples="examples/kafka/v1beta1/singlescramsecretassociation.yaml"

Brightside56 commented:
@turkenf , uptest has failed, 3.8.0 is non-existent Kafka version, I've replaced it with 3.8.x

jeanduplessis commented:
/test-examples="examples/kafka/v1beta1/singlescramsecretassociation.yaml"

mbbush commented:
/test-examples="examples/kafka/v1beta1/scramsecretassociation.yaml"

mbbush commented:
Now uptest is failing because of a mismatch in labels between the selector and the resource. Better to use `testing.upbound.io/example-name: single-scram-secret-association` everywhere.

One thing I'm surprised not to see in your manifest is the IAM resource policy for the secretsmanager secret. If you use the AWS Console to attach a secret to the MSK cluster, it applies the policy automatically, but that creates drift between the state in AWS and crossplane. I thought that I'd put the necessary policy in the `scramsecretassociation.yaml` manifest when I contributed that resource, but I don't see it now. If you'd be willing to add a [SecretPolicy](https://marketplace.upbound.io/providers/upbound/provider-aws-secretsmanager/v1.21.0/resources/secretsmanager.aws.upbound.io/SecretPolicy/v1beta1) resource for each secretsmanager secret in both your new manifest and the existing `scramsecretassociation.yaml` one, following the [policy](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/msk_scram_secret_association#example-usage) in the terraform provider docs (but in json), that would at a minimum provide better documentation for crossplane users, and hopefully also prevent another 2-hour long failed test run.

Brightside56 commented:
> Now uptest is failing because of a mismatch in labels between the selector and the resource. Better to use testing.upbound.io/example-name: single-scram-secret-association everywhere.

@mbbush Yes. I've made some typos

> One thing I'm surprised not to see in your manifest is the IAM resource policy for the secretsmanager secret. If you use the AWS Console to attach a secret to the MSK cluster, it applies the policy automatically, but that creates drift between the state in AWS and crossplane. I thought that I'd put the necessary policy in the scramsecretassociation.yaml manifest when I contributed that resource, but I don't see it now. If you'd be willing to add a [SecretPolicy](https://marketplace.upbound.io/providers/upbound/provider-aws-secretsmanager/v1.21.0/resources/secretsmanager.aws.upbound.io/SecretPolicy/v1beta1) resource for each secretsmanager secret in both your new manifest and the existing scramsecretassociation.yaml one, following the [policy](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/msk_scram_secret_association#example-usage) in the terraform provider docs (but in json), that would at a minimum provide better documentation for crossplane users, and hopefully also prevent another 2-hour long failed test run.

I've added policies, however I ask you to take a look on additional label on secrets (which is needed for selector in policies to get 1:1 mapping policy to secret if I understand correctly)

mbbush commented on examples/kafka/v1beta1/scramsecretassociation.yaml:
```diff
@@ -138,6 +140,70 @@ spec:
+        ]
+      }
+    region: us-east-2
+    secretArnSelector:
+      matchLabels:
+        testing.upbound.io/example-number: one
```
```suggestion
    secretArnRef:
      name: example1
```
Every field with a Selector also supports a Ref containing the metadata.name of a resource of the appropriate Kind. This way you don't need the extra labels.

Brightside56 replied:
Hm... It's a bit misleading, here I see secret id https://marketplace.upbound.io/providers/upbound/provider-aws-secretsmanager/v1.21.1/resources/secretsmanager.aws.upbound.io/SecretVersion/v1beta1, in secretpolicy I see secret arn... So, names of managed resources can be referenced and they will be resolved into arn, if I understand correct

Brightside56 replied:
I've replaced selector with secretArnRef in example

mbbush reviewed (commented):
Looking good, I'll run the tests.

mbbush commented:
/test-examples="examples/kafka/v1beta1/singlescramsecretassociation.yaml"

mbbush commented:
/test-examples="examples/kafka/v1beta1/scramsecretassociation.yaml"

mbbush reviewed (approved):
This LGTM, it just needs another round of codegen to populate the description update in `config` into the CRD and the comments in the `apis` package.

Brightside56 commented:
> This LGTM, it just needs another round of codegen to populate the description update in config into the CRD and the comments in the apis package.
Done

mbbush commented:
/test-examples="examples/kafka/v1beta1/singlescramsecretassociation.yaml"

mbbush commented:
Thank you @Brightside56! This should ship in the next release.
